packages/python/accern_data/accern_data.py

```diff
@@ -245,34 +265,66 @@ def max_date(
         # Second last date.
         return max(dates[:-1]).strftime(DT_FORMAT)
 
-    def do_init(self, is_first_day: bool) -> None:
+    def do_init(self, indicator: ProgressIndicator) -> None:
         fname = self.get_path(self._is_by_day)
-        if is_first_day or self._is_by_day:
+        if self._begining_of_file or self._is_by_day:
             pd.DataFrame([], columns=self._cols).to_csv(
                 fname, index=False, header=True, mode="w")
-        print_fn(f"current file is {fname}")
+        indicator.log(f"current file is {fname}")
 
-    def add_result(
-            self,
-            signal: Union[pd.DataFrame, List[Dict[str, Any]]]) -> None:
+    def add_result(self, signal: pd.DataFrame) -> None:
         fname = self.get_path(is_by_day=self._is_by_day)
-        assert isinstance(signal, pd.DataFrame)
         signal.to_csv(fname, index=False, header=False, mode="a")
 
-    def finish_day(self) -> None:
+    def finish_day(self, indicator: ProgressIndicator) -> None:
         # csv files are saved by add_result
-        pass
+        self._begining_of_file = self._is_by_day
 
     def split(
             self,
-            batch: Union[List[pd.DataFrame], List[Dict[str, Any]]],
-            value: pd.Timestamp) -> pd.DataFrame:
+            batch: List[pd.DataFrame],
+            value: pd.Timestamp) -> List[pd.DataFrame]:
         df = batch[0]
-        assert isinstance(df, pd.DataFrame)
-        return df[df["harvested_at"] <= value]
+        return [df[df["harvested_at"] <= value]]
 
-
-class JSONMode(Mode):
+    def iterate_data(
+            self,
+            data: Optional[List[pd.DataFrame]],
+            indicator: ProgressIndicator,
+            chunk_size: Optional[int] = None) -> Iterator[pd.DataFrame]:
+        assert chunk_size is None or chunk_size > 0
+        if chunk_size is None:
+            if data is not None:
+                df = data[0]
+                indicator.update(df.shape[0])
+                yield df
+        else:
+            if data is not None:
+                self._buffer.append(data[0])
+                self._buffer_size += data[0].shape[0]
+            while self._buffer_size >= chunk_size:
+                res_df = pd.concat(self._buffer)
+                result: pd.DataFrame = res_df.iloc[:chunk_size]
+                remainder: pd.DataFrame = res_df.iloc[chunk_size:]
+                if remainder.empty:
+                    self._buffer = []
+                else:
+                    self._buffer = [remainder]
+                self._buffer_size = remainder.shape[0]
+                indicator.update(result.shape[0])
+                yield result
```

Comment:
only yield non-empty dfs

Reply:
here, it's guaranteed that result is non-empty df.